### PR TITLE
cassandra_4: init

### LINKED ISSUE #167368: Cassandra 4.0
Cassandra 4.0 has been released. It would be great if we could supply it as `cassandra_4_0` initially, and eventually move to it as a default. See https://cassandra.apache.org/_/download.html

###### Checklist

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name

`nix search` name: `cassandra`

current version: 3.11.12
desired version: 4.0.3

###### Notify maintainers

maintainers: @roberth

###### Note for maintainers

Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by roberth**
###### Description of changes

This adds experimental cassandra 4 support. I have not tested a 3_11 -> 4 migration.

Closes #167368 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by turion** [comment]
Thanks a lot! Yeah, maybe it makes sense to add some tests that try to migrate, but I'm not sure how we would do that.

File: nixos/modules/services/databases/cassandra.nix [hunk 1]
```diff
@@ -94,8 +102,20 @@ let
...
+      # Historically, we don't use a log dir, whereas the upstream scripts do
+      # expect this. We override those by providing our own -Xlog:gc flag.
+      "-Xlog:gc=warning,heap*=warning,age*=warning,safepoint=warning,promotion*=warning"
     ];
 
+  commonEnv = {
+    # Sufficient for cassandra 2.x, 3.x
+    CASSANDRA_CONF = "${cassandraEtc}";
+
+    # Required since cassandra 4
```
[the remark this answers is not in the document]
Comment (roberth) [reply nixos/modules/services/databases/cassandra.nix:115]: The previous versions run, log and test just fine with. It is redundant but harmless, so let's keep it simple.

**Comment by roberth** [comment]
> I have not tested a 3_11 -> 4 migration.

This pr only adds the new package, and does not change any existing installations. More testing should be done when switching `pkgs.cassandra` to 4; a future pr.

**Comment by roberth** [comment]
> Yeah, maybe it makes sense to add some tests that try to migrate, but I'm not sure how we would do that.

A NixOS test could switch to a specialisation that has a newer version.

**Pull request merged**